internal/config/runtime.go

```diff
@@ -137,6 +142,51 @@ func (r *RuntimeConfig) GetContact(username string) *recipient.Contact {
 	return nil
 }
 
+// getSourceFromCredentialsUsernameRegexp is used in GetSourceFromCredentials to parse the HTTP basic auth username.
+//
+// Even when Source.ListenerPasswordHash is of type int64, the optional leading minus must not be parsed as the data
+// type within PostgreSQL, bigserial, is positive.
+var getSourceFromCredentialsUsernameRegexp = regexp.MustCompile(`source-(\d+)`)
+
+// GetSourceFromCredentials verifies a credential pair against known Sources.
+//
+// This method returns either a *Source or a nil pointer and logs the cause to the given logger. This is in almost all
+// cases a debug logging message, except when something server-side is wrong, e.g., the hash is invalid.
+func (r *RuntimeConfig) GetSourceFromCredentials(user, pass string, logger *logging.Logger) *Source {
+	r.RLock()
+	defer r.RUnlock()
+
+	usernameMatches := getSourceFromCredentialsUsernameRegexp.FindStringSubmatch(user)
+	if usernameMatches == nil {
+		logger.Debugw("Cannot extract source ID from HTTP basic auth username", zap.String("user-input", user))
+		return nil
+	}
```

Comment:
[`strings.CutPrefix()`](https://pkg.go.dev/strings#CutPrefix) should be much simpler here. Also, regarding negative numbers: it ends up as a map lookup, that doesn't care if `r.Sources[404]` or `r.Sources[-404]` is not within the map, so that shouldn't need any special treatment.